### PR TITLE
[Admin] Promotion features

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: features/promotion/managing_promotions/adding_promotion.feature [hunk 1]
```diff
@@ -16,3 +16,33 @@ Feature: Adding a new promotion
...
+        Given I want to create a new promotion
+        When I specify its code as "FULL_METAL_PROMOTION"
+        And I name it "Full metal promotion"
+        And I make it exclusive
+        And I add it
+        Then I should be notified that it has been successfully created
+        And the promotion "Full metal promotion" should be exclusive
+
+    @ui
+    Scenario: Adding a new exclusive promotion
```
Comment (Zales0123) [inline features/promotion/managing_promotions/adding_promotion.feature]: ``Adding a new coupon based promotion``

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php [hunk 2]
```diff
@@ -229,4 +283,15 @@ private function assertFieldValidationMessage($element, $expectedMessage)
...
+
+    /**
+     * @param PromotionInterface $promotion
+     * @param string $field
+     */
+    private function assertIfFieldIsTrue(PromotionInterface $promotion, $field)
+    {
+        $this->updatePage->open(['id' => $promotion->getId()]);
+
+        Assert::true($this->updatePage->hasResourceValues([$field => 1]), sprintf('Promotion %s is not exclusive, but it should be.', $promotion->getName()));
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php]: You have error message ``'Promotion %s is not exclusive, but it should be.'``, but this method is used also in ``thePromotionShouldBeCouponBased``.

File: src/Sylius/Bundle/AdminBundle/Resources/config/grids/promotion.yml [hunk 3]
```diff
@@ -4,21 +4,31 @@ sylius_grid:
             driver:
                 name: doctrine/orm
                 options:
-                    class: %sylius.model.promotion.class%
+                    class: "%sylius.model.promotion.class%"
             fields:
-                code:
-                    type: string
+                codeExlusiveAndCouponBased:
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/config/grids/promotion.yml]: ``codeExclusiveAndCouponBased``

File: src/Sylius/Bundle/AdminBundle/Resources/config/grids/promotion.yml [hunk 4]
```diff
@@ -4,21 +4,31 @@ sylius_grid:
...
+                    type: twig
                     label: sylius.ui.code
-                name:
-                    type: string
+                    path: .
+                    options:
+                        template: SyliusAdminBundle:Promotion/Field:codeExlusiveAndCouponBased.html.twig
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/config/grids/promotion.yml]: The same typo in template name.

File: src/Sylius/Bundle/AdminBundle/Resources/views/Promotion/Field/usedWithUsageLimit.html.twig [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<div class="ui red label">
+    {{ data.used }}
+    <div class="detail">
+        /
+        {% if data.usageLimit %}
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Promotion/Field/usedWithUsageLimit.html.twig]: Shouldn't this ``if`` be applied for all ``div.detail`` element?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Promotion/Field/usedWithUsageLimit.html.twig [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+                {{ data.usageLimit }}
+        {% else %}
+                ∞
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Promotion/Field/usedWithUsageLimit.html.twig]: Then we could avoid this ``else`` which is quite strange - I think it's obvious that promotion has no usage limit if it's not set :)

File: src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml [hunk 7]
```diff
@@ -35,6 +35,7 @@ sylius:
         filter: Filter
         in: In
         info: Info
+        promotion_configuration: Promotion configuration
```
Comment (Zales0123) [inline src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml]: Why here? :( Whole file is alphabetically ordered.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php [hunk 8]
```diff
@@ -229,4 +342,19 @@ private function assertFieldValidationMessage($element, $expectedMessage)
...
+     * @param PromotionInterface $promotion
+     * @param string $field
+     */
+    private function assertIfFieldIsTrue(PromotionInterface $promotion, $field)
+    {
+        $this->iWantToModifyAPromotion($promotion);
+
+        Assert::true(
+            $this->updatePage->hasResourceValues(
+                [$field => 1]),
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php]: Something has gone wrong here

File: src/Sylius/Behat/Page/Admin/Promotion/CreatePage.php [hunk 9]
```diff
@@ -79,6 +79,29 @@ public function fillActionOption($option, $value)
     /**
      * {@inheritdoc}
      */
+    public function fillUsageLimit($limit)
+    {
+        $this->getDocument()->fillField('Usage limit', $limit);
+    }
+
+    public function checkExclusive()
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Promotion/CreatePage.php]: makeExclusive?

File: src/Sylius/Behat/Page/Admin/Promotion/UpdatePage.php [hunk 10]
```diff
@@ -20,25 +20,21 @@
...
-        }
+        $field = $this->getDocument()->findField($channelName);
 
-        return true;
+        return (bool) $field->getValue();
     }
 
     /**
-     * {@inheritDoc}
+     * @inheritDoc
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Promotion/UpdatePage.php]: Why? :(

File: src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig [hunk 11]
```diff
@@ -72,14 +72,14 @@
...
-            <div data-form-collection="list"
-                 class="ui segments">
+            <div data-form-collection="list" >
                 {% for child in form %}
                     {{ _self.collectionItem(child, allow_delete, button_delete_label, loop.index0) }}
                 {% endfor %}
             </div>
 
             {% if prototype is defined and allow_add %}
+                <br/>
```
Comment (pamil) [inline src/Sylius/Bundle/UiBundle/Resources/views/Form/theme.html.twig]: Frontend experience++ 🐗

**Comment by pjedrzejewski** [comment]
Nice work Łukasz, thanks! 👍

**Pull request merged**